### ISSUE TITLE
[Epic] Testnet Automation

### ISSUE THREAD

**Issue by UnityChaos** [osmosis-labs/osmosis#656]
Tracking issue for automating testnet creation, upgrade, testing, etc.

The big things that are next step improvements seem to be:

- [x] Run pprof to see where memory is going during initialization
- [ ] Speedup InitGenesis/ExportGenesis time
- [x] Tendermint 2 peer required bug
- [ ] Testnets in CI

Future sprint (needs more details)
- [ ] Automate creation of export based testnets from Github Action
- [x] #732
- [ ] #729
- [ ] Github actions (using Terraform) for fully automated testnet creation and test running

**Comment by daniel-farina** [comment]
@UnityChaos @czarcas7ic 

I started a very simple example project with instructions on how to create a node in DigitalOcean with Terraform and then executing two of the scripts you created during the upgrade just as an example.  This is very basic and WIP but it should give you a good place to get started. 

- The server gets created (deleted if it already exist with the same name)
- Single validator testnet is created
- the genesis state is configured
- The other scripts I did not get to but this should give you a good idea of how things work.


This runs locally with Terraform, not via Terraform cloud. We will eventually have a scripts that configures Terraform cloud as well but I recommend getting familiar running Terraform locally first. 

Link to infrastructure repo: https://github.com/osmosis-labs/infrastructure

I think this infrastructure repo could grown into a collection of modules and different scenarios that includes front-end and backend. We can use it as we experiment with the tool.

**Comment by faddat** [comment]
I am not sure about the infra repo, because of:

#663 

and

https://github.com/osmosis-labs/infrastructure/issues/1

and of course, this isn't to say "don't do it" but maybe to say "let's do testnets as close as possible to how we do mainnets".

**Comment by czarcas7ic** [comment]
@UnityChaos @daniel-farina Would love to set up a call sometime to figure out the best way forward with this, specifically hashing out clear expectations for some of these bullets.

**Comment by ValarDragon** [comment]
This issue conjoins multiple things:
1) Starting up testnets more easily
2) Configurations for setting up testnets
3) Getting input & output from the testnet automatically
4) Getting testnet in CI completion steps

Everything depends on 1, and 1 is high priority. Adam's been doing a lot to improve it

I'm not sure what the plan/vision for 2 relative to 3 is

3 is important and a precursor for 4. Likely not a this-sprint priority though, as it is devops skill/time- constrained

**Comment by czarcas7ic** [comment]
@ValarDragon In my eyes, it seems like 1 and 2 are at a viable stage at the moment (i.e. we have an archive node that we can pull state exports from and then push them through the testnetify script in a short period of time). I feel the main focus should be 3, as figuring out all the test cases and expected results will be the tricky part. Once those tests are made, we can probably then circle back to using Terraform and making the entire process completely automated. 

Let me know if you believe there is a better way of breaking down this large task.

**Comment by UnityChaos** [comment]
Had a call with @czarcas7ic the other day to talk about some of this, and we came up some plans. I think you're right that points 3 and 4 may be worth delaying, relative to just being able to easily create export based testnets where we can manually test stuff.

We came up with a rough idea of how to do point 3 using python scripts. Basically just building some tooling for making calls to osmosisd to run transactions and queries, getting the results in json through STDOUT, and then asserting things about the expected results. It may be worth working on this independent of point 4 so that we can build up a robust set of tests to run, even if we have to run it manually. (I will create a separate issue for this and have it tracked here, to collapse a bunch of the todo items)

For point 4 I have a rough idea of what we want to do, but I'm not familiar enough with CI process to know how we will actually do it.
TLDR: Use Github Actions spin up a VM with Terraform/DO (or have a long lived kubernetes cluster we can deploy a new container on), then run the upgrade and test script and report results to CI. Fine with pushing this to the next sprint.

For point 1, I think with @czarcas7ic 's modification script (nice work on that) and what we have in the infrastructure repo, we're pretty close to being able to do this. Still need to work some of the process around getting a snapshot / export to modify, which I think could be somewhat complex to fully automate efficiently, but we could also just have a manual/inefficient process that we use infrequently. (ex: have a script/process for using chainlayer + export + modify with Packer to make a VM image with current mainnet version and a modified mainnet state, which we can use to spin up new VMs on demand to run upgrades and tests against)

For point 2, I think we should punt some of the potential complexity there by just focusing on export based testnets. Manually creating clean genesis testnets is already easy, but creating custom configurations is a pretty broad idea, and it might be better to just do that manually if/when we feel like we need it. (and automate later if we find that we're doing particular patterns a lot) (see also https://github.com/osmosis-labs/infrastructure/issues/3)

**Comment by czarcas7ic** [comment]
@UnityChaos We posted at the same time but looks like we were pretty lock-step in agreement here! I know you and @ValarDragon have a lot of experience with doing these manually from past exports, so going through this and documenting expected outputs seems like the logical next step.

**Comment by UnityChaos** [comment]
Yup lol. Gonna clean up the todo items a bit and make a separate issue for part 3 and then we can start listing out some expected outputs for various test patterns :+1:

**Comment by ValarDragon** [comment]
The big things that are next step improvements seem to be:
- [ ] Speedup InitGenesis/ExportGenesis time
- [ ] Testnets in CI

**Comment by czarcas7ic** [comment]
![stateexportinportheap](https://user-images.githubusercontent.com/40078083/158894329-0cb1aca1-a92c-49ef-a2c3-64f7bcc76e92.png)

**Comment by czarcas7ic** [comment]
![stateexportimportcpu](https://user-images.githubusercontent.com/40078083/158894408-9f5a7a49-62be-414a-940d-d03ce1386e0e.png)

**Comment by czarcas7ic** [comment]
As discussed previously, I have posted the pprof output taken right after initgen is completed on a new state exported testnet

**Comment by czarcas7ic** [comment]
I switched out the decoding/json package for goccy/go-json which is supposed to be much faster, here are the pprof results (same machine, time frame, and genesis)


![profile002](https://user-images.githubusercontent.com/40078083/159285938-0007d80d-c821-48d3-89e9-b605a122d3ad.png)


![profile001](https://user-images.githubusercontent.com/40078083/159286125-327e9a3c-84b2-442f-a4ff-fe8e4ddbc88e.png)

**Comment by ValarDragon** [comment]
Closing in favor of #1181
